tools/state-viewer/src/apply_chain_range.rs

```diff
@@ -353,59 +364,87 @@ pub fn apply_chain_range(
+        }
+    }
+
+    let (start_height, end_height) = match (mode, storage) {
+        (ApplyRangeMode::Benchmark, StorageSource::Trie | StorageSource::TrieFree) => {
+            panic!("benchmark with --storage trie|trie-free is not supported")
+        }
+        (ApplyRangeMode::Benchmark, StorageSource::FlatStorage | StorageSource::Memtrie) => {
+            // Benchmarking mode requires flat storage and retrieves the block height from flat
+            // storage.
             assert!(start_height.is_none());
             assert!(end_height.is_none());
-
-            let chain_store =
-                ChainStore::new(read_store.clone(), genesis.config.genesis_height, false);
-            let final_head = chain_store.final_head().unwrap();
-            let shard_layout = epoch_manager.get_shard_layout(&final_head.epoch_id).unwrap();
-            let shard_uid = near_primitives::shard_layout::ShardUId::from_shard_id_and_layout(
-                shard_id,
-                &shard_layout,
-            );
             let flat_head = match read_store.flat_store().get_flat_storage_status(shard_uid) {
                 Ok(FlatStorageStatus::Ready(ready_status)) => ready_status.flat_head,
                 status => {
-                    panic!("cannot create flat storage for shard {shard_id} with status {status:?}")
+                    panic!(
+                        "cannot create flat storage for shard {shard_uid} with status {status:?}"
+                    )
                 }
             };
-            let flat_storage_manager = runtime_adapter.get_flat_storage_manager();
-            flat_storage_manager.create_flat_storage_for_shard(shard_uid).unwrap();
-
-            // Note that first height to apply is the first one after flat
-            // head.
-            (flat_head.height + 1, final_head.height)
+            // We apply the block at flat_head. Users can set the block they want to benchmark by
+            // moving the flat head using the `flat-storage move-flat-head` command.
+            (flat_head.height + 1, 0)
         }
-        _ => (
+        (_, StorageSource::Trie | StorageSource::TrieFree) => (
             start_height.unwrap_or_else(|| chain_store.tail().unwrap()),
             end_height.unwrap_or_else(|| chain_store.head().unwrap().height),
         ),
+        (_, StorageSource::FlatStorage | StorageSource::Memtrie) => {
+            (
+                start_height.unwrap_or_else(|| {
+                    let flat_head = match read_store.flat_store().get_flat_storage_status(shard_uid) {
+                        Ok(FlatStorageStatus::Ready(ready_status)) => ready_status.flat_head,
+                        status => {
+                            panic!(
+                                "cannot create flat storage for shard {shard_uid} with status {status:?}"
+                            )
+                        }
+                    };
+                    flat_head.height + 1
```

Comment:
This seems duplicated, can you move it to a helper method?

Reply:
Seems difficult to abstract out into an useful method. Closest I got to something sensible is

```rust
    pub fn get_flat_head(
        &self,
        shard_uid: ShardUId,
    ) -> Result<Option<BlockInfo>, FlatStorageError> {
        Ok(match self.get_flat_storage_status(shard_uid)? {
            FlatStorageStatus::Ready(ready_status) => Some(ready_status.flat_head),
            _ => None
        })
    }
```

but this loses the Ok(non-Ready) status information which is currently printed out on panic within the tool.